### DISCUSSION TITLE
OpenFGA support

### DISCUSSION THREAD

**Discussion by Kremr** [zitadel/zitadel#5729] [Ideas]
As a SaaS Solution Architect I need besides the authentication logic also a system for permission checks. One solution which would perfectly fit could be https://github.com/openfga/openfga. List of adopters of OpenFGA: https://github.com/openfga/community/blob/main/ADOPTERS.md

"OpenFGA is an open-source authorization solution that allows developers to build granular access control using an easy-to-read modeling language and friendly APIs."

Info links and other solutions:
https://research.google/pubs/pub48190/
https://en.wikipedia.org/wiki/Relationship-based_access_control
https://www.topaz.sh
https://warrant.dev

**Comment by ghstahl** [comment]
I use OpenFGA and plan to deploy Zitadel as well.  At the moment I see no need for any direct integration between zitadel and my openFGA.  The biggest reason is that my use of OpenFGA is customized to my company.  My idea of a user and organization on the OpenFGA side is not the same as what zitadels idea of an organization is and what relationship that user has to any org.

I am planning to have a custom zitadel action (post create) that makes a call into my OpenFGA to add the zitadel.user.subject to my openfga graph.  That is the extent of the link between the two systems.
I am also leaning toward NOT having a custom action to call into our systems at user create time.  Instead, our webapps will make a EnsureUser() call to our openFGA micro (passing the users access_token) and it will add the user to openFGA.  

My OpenFGA model is where all permissions are.  The use of zitadel is only "Proof of Identity".  Very simple access_token(just need the subject).    After that it's all OpenFGA calls. 

Maybe a token_exchange call zitadel.user.access_token + OpenFGA(users permissions in context of org) == the final access_token which is now full of claims. 

 I would like to see other design proposals that have a direct integration.

**Reply by danielloader** [reply]
@ghstahl How did this go? Did you find a benefit to having two API calls (and the problems if one/both fail) when creating a user?

We are also debating the pros and cons of utilising a hook in the actions v2 model to trigger an addition of a tuple to an OpenFGA service at creation time, or have the frontend speak to a monolithic backend which handles both calls to zitadel and openfga in sequence, or if the frontend should talk to zitadel and openFGA (exposed from kubernetes) itself.

**Reply by johnnypea** [reply]
@ghstahl @danielloader Which implementation did you choose? Thanks.
